File: AirPurrr/app/src/main/java/com/example/trubul/airpurrr/SwitchHelper.java
package com.example.trubul.airpurrr;

import android.content.Context;
import android.widget.CompoundButton;
import android.widget.Toast;

import java.util.concurrent.ExecutionException;

class SwitchHelper implements CompoundButton.OnCheckedChangeListener {
    private static final String WORKSTATE_URL = "http://airpurrr.ga/workstate.txt";
    private Context mContext;
    private SwitchCallback mCallback;
    private boolean stateManual = false;

    interface SwitchCallback {
        void setSwitchManual(boolean state);
    }

    SwitchHelper(Context context, SwitchCallback callback) {
        mContext = context;
        mCallback = callback;
    }

    @Override
    public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
        stateManual = isChecked;
        controlRequests(stateManual);
    }

    private void controlRequests(boolean state) {
        String workStates;
        HttpsPostRequest switchOn = new HttpsPostRequest();
        HttpsPostRequest switchOff = new HttpsPostRequest();

        try {
            HttpGetRequest getRequest = new HttpGetRequest();
            workStates = getRequest.execute(WORKSTATE_URL).get();

            switch (workStates) {
                case "WorkStates.Sleeping\n":
                    Toast.makeText(mContext, R.string.main_message_switch_processing, Toast.LENGTH_LONG).show();
                    if (state) {  // send request if it was switch -> ON
                        switchOn.execute("MANUAL=1");  // it will be POST: req = params[0]

                    } else {
                        switchOff.execute("MANUAL=0");
                    }
                    break;
                case "WorkStates.Measuring\n":
                    Toast.makeText(mContext, R.string.main_message_error_measuring, Toast.LENGTH_LONG).show();
                    keepState();
                    break;
                default:
                    Toast.makeText(mContext, R.string.main_message_error + "NoWorkStates)", Toast.LENGTH_LONG).show();
                    keepState();
                    break;
            }
        } catch (InterruptedException | ExecutionException e) {
            e.printStackTrace();
        } catch (NullPointerException e) {
            Toast.makeText(mContext, mContext.getString(R.string.main_message_error_server) + "NullPointer)", Toast.LENGTH_LONG).show();
            keepState();
        }
    }

    private void keepState() {
        stateManual = !stateManual;
        mCallback.setSwitchManual(stateManual);
    }

}

File: AirPurrr/app/src/main/java/com/example/trubul/airpurrr/SwitchHelper.kt
package com.example.trubul.airpurrr

import android.view.View
import android.widget.CompoundButton
import com.example.trubul.airpurrr.model.Detector
import com.example.trubul.airpurrr.retrofit.DetectorService
import com.google.android.material.snackbar.Snackbar
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

import java.util.concurrent.ExecutionException

internal class SwitchHelper(private val mParentLayout: View, private val hashedEmail: String,
                            private val hashedPassword: String, private val mCallback: SwitchCallback) : CompoundButton.OnCheckedChangeListener {

    private var stateManual = false
    var workStates = ""

    internal interface SwitchCallback {
        fun setSwitchManual(state: Boolean)
    }

    override fun onCheckedChanged(buttonView: CompoundButton, isChecked: Boolean) {
        stateManual = isChecked
        retrofitSwitch()
    }

    private fun controlRequests(state: Boolean) {
        val switchOn = HttpsPostRequest(hashedEmail, hashedPassword)
        val switchOff = HttpsPostRequest(hashedEmail, hashedPassword)

        try {
            when (workStates) {
                "WorkStates.Sleeping" -> {
                    Snackbar.make(mParentLayout, R.string.main_message_switch_processing, Snackbar.LENGTH_LONG).show()
                    if (state) {  // send request if it was switch -> ON
                        switchOn.execute("MANUAL=1")  // it will be POST: req = params[0]

                    } else {
                        switchOff.execute("MANUAL=0")
                    }
                }
                "WorkStates.Measuring" -> {
                    Snackbar.make(mParentLayout, R.string.main_message_error_measuring, Snackbar.LENGTH_LONG).show()
                    keepState()
                }
                else -> {
                    Snackbar.make(mParentLayout, R.string.main_message_error, Snackbar.LENGTH_LONG).show()
                    keepState()
                }
            }
        } catch (e: InterruptedException) {
            e.printStackTrace()
        } catch (e: ExecutionException) {
            e.printStackTrace()
        } catch (e: NullPointerException) {
            Snackbar.make(mParentLayout, R.string.main_message_error_server, Snackbar.LENGTH_LONG).show()
            keepState()
        }

    }

    private fun keepState() {
        stateManual = !stateManual
        mCallback.setSwitchManual(stateManual)
    }

    fun retrofitSwitch() {
        val service by lazy { DetectorService.create() }

        val call = service.getDetectorData()
        call.enqueue(object : Callback<Detector.Result> {
            override fun onResponse(call: Call<Detector.Result>, response: Response<Detector.Result>) {
                val data = response.body()
                if (data != null) {
                    workStates = data.workstate
                    controlRequests(stateManual)
                }
            }

            override fun onFailure(call: Call<Detector.Result>, t: Throwable) {
                workStates = "error"
            }
        })
    }
}